nixos/modules/virtualisation/qemu-vm.nix

```diff
@@ -135,32 +135,32 @@ let
           TMPDIR=$(mktemp -d nix-vm.XXXXXXXXXX --tmpdir)
       fi
 
-      ${lib.optionalString (cfg.useNixStoreImage)
-        (if cfg.writableStore
-          then ''
-            # Create a writable copy/snapshot of the store image.
-            ${qemu}/bin/qemu-img create -f qcow2 -F qcow2 -b ${storeImage}/nixos.qcow2 "$TMPDIR"/store.img
-          ''
-          else ''
-            (
-              cd ${builtins.storeDir}
-              ${hostPkgs.erofs-utils}/bin/mkfs.erofs \
-                --force-uid=0 \
-                --force-gid=0 \
-                -L ${nixStoreFilesystemLabel} \
-                -U eb176051-bd15-49b7-9e6b-462e0b467019 \
-                -T 0 \
-                --exclude-regex="$(
-                  <${hostPkgs.closureInfo { rootPaths = [ config.system.build.toplevel regInfo ]; }}/store-paths \
-                    sed -e 's^.*/^^g' \
-                  | cut -c -10 \
-                  | ${hostPkgs.python3}/bin/python ${./includes-to-excludes.py} )" \
-                "$TMPDIR"/store.img \
-                . \
-                </dev/null >/dev/null
-            )
-          ''
-        )
+      ${lib.optionalString (cfg.useNixStoreImage) ''
+        echo "Creating Nix store image..."
+
+        # Keep this around only as long as necessary to create the filesystem
+        # image.
+        nixStorePaths="$TMPDIR"/nix-store-paths
+        mkdir -p "$nixStorePaths"
+
+        <${hostPkgs.closureInfo { rootPaths = [ config.system.build.toplevel regInfo ]; }}/store-paths \
+          xargs cp --archive --target-directory "$nixStorePaths"
```

Comment:
`mkfs.erofs --tar` should be able to read from a tarball, which means that we should be able to control the contents quite well without having to copy the store paths in advance.
Have you tried that? It could be faster.

Reply:
Indeed building a tar seems to be a little faster. On my machine running the test now takes 11s vs 13s before. This gets it much closer to the old regex times (10s).